File: MERNstack/frontend/src/lib/auth-service.ts
// Authentication service connected to Express backend
const API_URL = "https://lamp-stack4331.xyz/api";

export interface User {
  _id: string;
  email: string;
  name: string;
  phone?: string;
  balance: number;
}

export interface AuthResponse {
  user: User;
  token: string;
}

export const authService = {
  async login(email: string, password: string): Promise<AuthResponse> {
    const response = await fetch(`${API_URL}/auth/login`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, password }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Invalid email or password");
    }

    const data = await response.json();
    this.storeToken(data.token);
    this.storeUser(data.user);
    return data;
  },

  async signup(
    email: string,
    password: string,
    name: string
  ): Promise<AuthResponse> {
    const response = await fetch(`${API_URL}/auth/register`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email, password, name }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Signup failed");
    }

    const data = await response.json();
    this.storeToken(data.token);
    this.storeUser(data.user);
    return data;
  },

  async resetPassword(email: string): Promise<{ message: string }> {
    const response = await fetch(`${API_URL}/auth/reset-password`, {
      method: "POST",
      headers: { "Content-Type": "application/json" },
      body: JSON.stringify({ email }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Password reset failed");
    }

    return response.json();
  },

  async updateUser(updatedUser: Partial<User>): Promise<User> {
    const token = this.getStoredToken();
    if (!token) throw new Error("Not authenticated");

    const response = await fetch(`${API_URL}/users/update`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify(updatedUser),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Update failed");
    }

    const data = await response.json();
    this.storeUser(data.user);
    return data.user;
  },

  async addBalance(amount: number): Promise<User> {
    const token = this.getStoredToken();
    if (!token) throw new Error("Not authenticated");

    const response = await fetch(`${API_URL}/users/addBalance`, {
      method: "POST",
      headers: {
        "Content-Type": "application/json",
        Authorization: `Bearer ${token}`,
      },
      body: JSON.stringify({ amount }),
    });

    if (!response.ok) {
      const error = await response.json();
      throw new Error(error.error || "Failed to add balance");
    }

    const data = await response.json();
    const updatedUser = await this.getUser();
    if (!updatedUser) throw new Error("Failed to fetch updated user");
    return updatedUser;
  },

  getUser(): User | null {
    const user = localStorage.getItem("user");
    return user ? JSON.parse(user) : null;
  },

  storeUser(user: User): void {
    localStorage.setItem("user", JSON.stringify(user));
  },

  clearUser(): void {
    localStorage.removeItem("user");
  },

  logout(): void {
    localStorage.removeItem("auth_token");
    localStorage.removeItem("user");
  },

  getStoredToken(): string | null {
    return localStorage.getItem("auth_token");
  },

  storeToken(token: string): void {
    localStorage.setItem("auth_token", token);
  },
};

File: MERNstack/mobile/big_project_mobile_app/lib/services/auth_service.dart
// lib/services/auth_service.dart

import 'dart:convert';
import 'package:http/http.dart' as http;
import '../api_config.dart';

class AuthService {
  // Simple global store for demo purposes
  static String? authToken;
  static Map<String, dynamic>? currentUser;

  /// Logs in using /api/auth/login and stores token + user.
  Future<Map<String, dynamic>> login(String email, String password) async {
    final uri = Uri.parse('$apiBaseUrl/api/auth/login');

    final res = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode({'email': email, 'password': password}),
    );

    final data = _safeJson(res.body);
    if (res.statusCode == 200) {
      authToken = data['token'];
      currentUser = data['user'];
      return {'success': true, 'token': data['token'], 'user': data['user']};
    } else {
      return {'success': false, 'message': data['error'] ?? 'Login failed'};
    }
  }

  /// Adds $100 demo funds using /api/users/addBalance, then fetches /api/users/me.
  Future<Map<String, dynamic>> addDemoFunds() async {
    if (authToken == null) {
      return {'success': false, 'message': 'You must be logged in to add funds'};
    }

    final addUri = Uri.parse('$apiBaseUrl/api/users/addBalance');
    final addRes = await http.post(
      addUri,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $authToken',
      },
    );

    if (addRes.statusCode != 200) {
      final data = _safeJson(addRes.body);
      return {'success': false, 'message': data['error'] ?? 'Failed to add funds'};
    }

    final meUri = Uri.parse('$apiBaseUrl/api/users/me');
    final meRes = await http.post(
      meUri,
      headers: {
        'Content-Type': 'application/json',
        'Authorization': 'Bearer $authToken',
      },
    );

    if (meRes.statusCode == 200) {
      final data = _safeJson(meRes.body);
      final user = (data['user'] as Map<String, dynamic>? ?? {});
      currentUser = user;
      final rawBalance = user['balance'];
      final balance = (rawBalance is num) ? rawBalance.toDouble() : 0.0;
      return {'success': true, 'balance': balance};
    }
    return {'success': true};
  }

  /// Registers a new user and stores token + user.
  /// Adjust the endpoint path if your backend uses '/api/auth/signup' instead.
  Future<Map<String, dynamic>> register({
    required String email,
    required String password,
    String? name,
  }) async {
    final uri = Uri.parse('$apiBaseUrl/api/auth/register'); // change to /signup if needed

    final body = {
      'email': email,
      'password': password,
      if (name != null && name.trim().isNotEmpty) 'name': name.trim(),
    };

    final res = await http.post(
      uri,
      headers: {'Content-Type': 'application/json'},
      body: jsonEncode(body),
    );

    final data = _safeJson(res.body);
    if (res.statusCode == 201 || res.statusCode == 200) {
      authToken = data['token'];
      currentUser = data['user'];
      return {'success': true, 'token': data['token'], 'user': data['user']};
    } else {
      return {'success': false, 'message': data['error'] ?? 'Registration failed'};
    }
  }

  Map<String, dynamic> _safeJson(String body) {
    try {
      final v = jsonDecode(body);
      return (v is Map<String, dynamic>) ? v : <String, dynamic>{};
    } catch (_) {
      return <String, dynamic>{};
    }
  }
}
